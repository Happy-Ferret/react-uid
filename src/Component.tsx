import * as React from 'react';

let counter = {
  value: 0
};

const context = React.createContext({value: 0});

export interface UIDProps {
  name?: (n: number) => string,
  idSource?: { value: number },
  children: (n: string) => React.ReactNode
}

const getId = (source: typeof counter) => source.value++;

export class UID extends React.Component<UIDProps> {
  state = {
    id: getId(this.props.idSource || counter)
  };

  render() {
    const {name, children} = this.props;
    const {id} = this.state;
    return children(String(name ? name(id) : id))
  }
}

export const UIDReset: React.SFC = ({children}) => <context.Provider value={{value: 0}}>{children}</context.Provider>;
export const SmartUID: React.SFC<UIDProps> = (props) => (
  <context.Consumer>
    {
      (value) => <UID {...props} idSource={value}/>
    }
  </context.Consumer>
);