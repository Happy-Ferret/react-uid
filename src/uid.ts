export const generateUID = () => {
  let counter = 1;

  const map = new Map<any, number>();

  const uid = (item: any): string => {
    if (
      typeof item === 'number' ||
      typeof item === 'string'
    ) {
      return 'not-uid' + String(item);
    }

    if (!map.has(item)) {
      map.set(item, counter++)
      return uid(item)
    }
    return 'uid' + map.get(item);
  };

  return uid;
}