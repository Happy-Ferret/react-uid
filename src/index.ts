import {generateUID} from "./uid";
import {UID, UIDReset, SmartUID} from "./Component";

const uid = generateUID();

export {
  generateUID,
  uid,
  UID,
  UIDReset,
  SmartUID
}