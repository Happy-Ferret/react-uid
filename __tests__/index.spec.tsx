import * as React from 'react';
import * as TestRenderer from 'react-test-renderer';
import {uid, UID, UIDReset, generateUID, SmartUID} from "../src";


describe('uid', () => {
  it('test uid', () => {
    const a = {};
    const b = {}
    expect(uid(a)).toEqual('uid1');
    expect(uid(a)).toEqual('uid1');
    expect(uid(b)).toEqual('uid2');
    expect(uid(a)).toEqual('uid1');

    expect(uid("test")).toEqual("not-uidtest");
    expect(uid(42)).toEqual("not-uid42");
  });

  it('generate', () => {
    const uid1 = generateUID();
    const uid2 = generateUID();

    const a = {};
    const b = {}
    expect(uid1(a)).toEqual('uid1');
    expect(uid2(b)).toEqual('uid1');
    expect(uid1(b)).toEqual('uid2');
    expect(uid2(a)).toEqual('uid2');
  })
});

describe('UID', () => {
  it('test UID', () => {
    const testRenderer = TestRenderer.create(
      <div>
        <UID>
          {id => <span>{id}</span>}
        </UID>
        <UID>
          {id => <span>{id}</span>}
        </UID>
      </div>
    );

    testRenderer.update(
      <div>
        <UID>
          {id => <span>{id}</span>}
        </UID>
        <UID>
          {id => <span>{id}</span>}
        </UID>
      </div>
    );

    expect(testRenderer.toJSON()).toMatchSnapshot();

    const testRenderer2 = TestRenderer.create(
      <div>
        <UID name={id => `key-${id}`}>
          {id => <span>{id}</span>}
        </UID>
      </div>
    );

    expect(testRenderer2.toJSON()).toMatchSnapshot();
  })

  it('test UID', () => {
    const testRenderer1 = TestRenderer.create(
      <UIDReset>
        <SmartUID>
          {id => <span>{id}</span>}
        </SmartUID>
        <SmartUID>
          {id => <span>{id}</span>}
        </SmartUID>
        <UID>
          {id => <span>{id}</span>}
        </UID>
      </UIDReset>
    );

    const testRenderer2 = TestRenderer.create(
      <UIDReset>
        <SmartUID>
          {id => <span>{id}</span>}
        </SmartUID>
        <UIDReset>
          <SmartUID>
            {id => <span>{id}</span>}
          </SmartUID>
        </UIDReset>
        <UID>
          {id => <span>{id}</span>}
        </UID>
      </UIDReset>
    );


    expect(testRenderer1.toJSON()).toMatchSnapshot();
    expect(testRenderer2.toJSON()).toMatchSnapshot();
  })
});
