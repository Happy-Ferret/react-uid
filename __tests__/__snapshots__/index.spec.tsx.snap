// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UID test UID 1`] = `
<div>
  <span>
    0
  </span>
  <span>
    1
  </span>
</div>
`;

exports[`UID test UID 2`] = `
<div>
  <span>
    key-2
  </span>
</div>
`;

exports[`UID test UID 3`] = `
Array [
  <span>
    0
  </span>,
  <span>
    1
  </span>,
  <span>
    3
  </span>,
]
`;

exports[`UID test UID 4`] = `
Array [
  <span>
    0
  </span>,
  <span>
    0
  </span>,
  <span>
    4
  </span>,
]
`;
